src/value_log.rs

```diff
@@ -0,0 +1,374 @@
+    /// `write` should not be called concurrently, otherwise this will lead to wrong result.
+    pub fn write(&self, requests: &mut [Request]) -> Result<()> {
+        let result = self.write_inner(requests);
+        if self.opts.sync_writes {
+            let core = self.core.read().unwrap();
+            let current_log_id = core.max_fid;
+            let current_log_ptr = core.files_map.get(&current_log_id).unwrap().clone();
+            let mut current_log = current_log_ptr.write().unwrap();
+            drop(core);
+            current_log.sync()?;
+        }
+        result
+    }
+
+    pub fn write_inner(&self, requests: &mut [Request]) -> Result<()> {
+        let core = self.core.read().unwrap();
+        let mut current_log_id = core.max_fid;
+        let mut current_log = core.files_map.get(&current_log_id).unwrap().clone();
+        drop(core);
+
+        // `write` is called serially. There won't be two routines concurrently
+        // calling this function. Therefore, we could bypass a lot of lock schemes
+        // in this function.
+        let write = |buf: &[u8], current_log_lck: &RwLock<Wal>| -> Result<()> {
+            if buf.is_empty() {
+                return Ok(());
+            }
+            let n = buf.len() as u32;
+            let start = self
+                .writeable_log_offset
+                .fetch_add(n, std::sync::atomic::Ordering::SeqCst);
+            let end_offset = start + n;
+
+            // expand file size if space is not enough
+            // TODO: handle value >= 4GB case
+            let mut current_log = current_log_lck.write().unwrap();
+            if end_offset >= current_log.size() {
+                current_log.set_len(end_offset as u64)?;
+            }
+            // As `start..end_offset` is only used by current write routine, we
+            // could safely unlock the log lock and copy data inside.
+            let ptr = current_log.data()[start as usize..end_offset as usize].as_mut_ptr();
+            drop(current_log);
+            unsafe {
+                std::ptr::copy_nonoverlapping(buf.as_ptr(), ptr, buf.len());
+            }
+            // ensure data are flushed to main memory
+            std::sync::atomic::fence(std::sync::atomic::Ordering::SeqCst);
+
+            let mut current_log = current_log_lck.write().unwrap();
+            current_log.set_size(end_offset);
+            drop(current_log);
+            Ok(())
+        };
+
+        // `to_disk` returns `true` if we need a new vLog.
+        let to_disk = |current_log: &RwLock<Wal>| -> Result<bool> {
+            let core = self.core.read().unwrap();
+            if self.w_offset() as u64 > self.opts.value_log_file_size
+                || core.num_entries_written > self.opts.value_log_max_entries
+            {
+                let mut current_log = current_log.write().unwrap();
+                current_log.done_writing(self.w_offset())?;
```

Comment:
This seems can race with L181, is it safe?

Reply:
All sub-functions in this `write_inner` will not be spawned to other threads. They are all running in one thread. Things like `to_disk` and `write` are used many times throughout the vLog writing process, so they are made into two separate sub-routines.

Reply:
In fact, I really think Badger needs refactoring this part...